rfcs/0000-import-expression-ast-node.md

```diff
@@ -0,0 +1,251 @@
+- Repo: `babel/babel`
+- Start Date: 2020-05-01
+- RFC PR: https://github.com/babel/rfcs/pull/4
+- Related Issues: https://github.com/babel/babel/pull/10962
+- Authors: @JLHwung
+- Champion: @JLHwung
+- Implementors: @JLHwung
+
+# Summary
+
+This RFC proposes to change the AST shape of dynamic import `import(moduleName)` in ES2020 by introducing a new `ImportExpression` node.
+
+# Motivation
+
+Currently dynamic import is parsed as a `CallExpression`, whose `callee` is a pseudo expression node `Import`. For example, here is the parsed AST of `import("foo")`.
+
+```jsonc
+// import("foo")
+{
+  "type": "CallExpression",
+  "callee": {
+    "type": "Import"
+  },
+  "arguments": [
+    {
+      "type": "StringLiteral",
+      "value": "foo"
+    }
+  ]
+}
+```
+
+When `@babel/parser` [supported](https://github.com/babel/babylon/pull/163) dynamic import, it was called `import-function` and shared similar syntax structure with `super` call that is parsed as `CallExpression[callee=Super]`. However, it encourages an incorrect mental model, as developers may think of `import()` as a function call. Besides, it was later syntactically limited to one argument, different from other call expressions. Last, it is also problematic when we extend the current AST shape to support stage-1 [module attributes](https://github.com/tc39/proposal-module-attributes) (`import(moduleName, attributes)`) since the arguments (`moduleName` and `attributes` respectively) share the same semantics in the AST.
```

Comment:
There's a bit of discussion in https://github.com/tc39/proposal-module-attributes/issues/52. Honestly, I'd rather we treat it as a regular function call, exactly the same as `super()`.

Reply:
I very much doubt that `import()` will ever get consensus to be treated as a normal function call; there was a lot of discussion about that during its advancement. It's syntax and will likely always be special in a number of ways.

Reply:
Note that the `ImportExpression` is also created to be consistent with `ImportDeclaration`. 

```js
interface ImportDeclaration <: ModuleDeclaration {
  type: "ImportDeclaration";
  source: Literal;
}

interface ImportExpression <: Expression {
  type: "ImportExpression";
  source: Expression;
}
```

Although on the CST level an import expression shares similar syntax features with a function call, but on the AST level I think we should abstract them to common import interface.